// taco-dependency-installer/installers/androidSdkInstaller.js
/**
  *******************************************************
  *                                                     *
  *   Copyright (C) Microsoft. All rights reserved.     *
  *                                                     *
  *******************************************************
  */
/// <reference path="../../typings/adm-zip.d.ts" />
/// <reference path="../../typings/dependencyInstallerInterfaces.d.ts" />
/// <reference path="../../typings/Q.d.ts" />
/// <reference path="../../typings/request.d.ts" />
/// <reference path="../../typings/wrench.d.ts" />
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var AdmZip = require("adm-zip");
var childProcess = require("child_process");
var fs = require("fs");
var os = require("os");
var path = require("path");
var Q = require("q");
var util = require("util");
var wrench = require("wrench");
var InstallerBase = require("./installerBase");
var installerUtils = require("../utils/installerUtils");
var installerUtilsWin32 = require("../utils/win32/installerUtilsWin32");
var resources = require("../resources/resourceManager");
var tacoUtils = require("taco-utils");
var utilHelper = tacoUtils.UtilHelper;
var AndroidSdkInstaller = (function (_super) {
    __extends(AndroidSdkInstaller, _super);
    function AndroidSdkInstaller(installerInfo, softwareVersion, installTo, logger, steps) {
        _super.call(this, installerInfo, softwareVersion, installTo, logger, steps, "androidSdk");
    }
    AndroidSdkInstaller.prototype.downloadWin32 = function () {
        return this.downloadDefault();
    };
    AndroidSdkInstaller.prototype.installWin32 = function () {
        return this.installDefault();
    };
    AndroidSdkInstaller.prototype.updateVariablesWin32 = function () {
        // Initialize values
        var androidHomeValue = path.join(this.installDestination, "android-sdk-windows");
        var addToPathTools = path.join(androidHomeValue, "tools");
        var addToPathPlatformTools = path.join(androidHomeValue, "platform-tools");
        this.androidHomeValue = androidHomeValue;
        return installerUtilsWin32.setEnvironmentVariableIfNeededWin32(AndroidSdkInstaller.ANDROID_HOME_NAME, androidHomeValue, this.logger)
            .then(function () {
            return installerUtilsWin32.addToPathIfNeededWin32([addToPathTools, addToPathPlatformTools]);
        });
    };
    AndroidSdkInstaller.prototype.downloadDarwin = function () {
        return this.downloadDefault();
    };
    AndroidSdkInstaller.prototype.installDarwin = function () {
        var self = this;
        // Before we extract Android SDK, we need to save the first directory under the specified install path that doesn't exist. This directory and all those under it will be created
        // with root as the owner, so we will need to change the owner back to the current user after the extraction is complete.
        var pathSegments = path.resolve(this.installDestination).split(os.EOL);
        var firstNonExistentDir;
        var pathSoFar = "";
        pathSegments.some(function (dir) {
            pathSoFar = path.join(pathSoFar, dir);
            if (!fs.existsSync(pathSoFar)) {
                firstNonExistentDir = pathSoFar;
                return true;
            }
            return false;
        });
        return this.installDefault()
            .then(function () {
            // If some segments of the path the SDK was extracted to didn't exist before, it means they were created as part of the install. They will have root as the owner, so we 
            // must change the owner back to the current user.
            if (firstNonExistentDir) {
                wrench.chownSyncRecursive(firstNonExistentDir, parseInt(tacoUtils.ProcessUtils.getProcess().env.SUDO_UID, 10), parseInt(tacoUtils.ProcessUtils.getProcess().env.SUDO_GID, 10));
            }
        });
    };
    AndroidSdkInstaller.prototype.updateVariablesDarwin = function () {
        var _this = this;
        var deferred = Q.defer();
        // Initialize values
        var androidHomeValue = path.join(this.installDestination, "android-sdk-macosx");
        var fullPathTools = path.join(androidHomeValue, "tools/");
        var fullPathPlatformTools = path.join(androidHomeValue, "platform-tools/");
        var shortPathTools = util.format("$%s%stools", AndroidSdkInstaller.ANDROID_HOME_NAME, path.sep);
        var shortPathPlatformTools = util.format("$%s%splatform-tools", AndroidSdkInstaller.ANDROID_HOME_NAME, path.sep);
        var addToPath = "";
        var exportPathLine = "";
        var exportAndroidHomeLine = "";
        var updateCommand = "";
        var useShortPaths = true;
        // Save Android home value
        this.androidHomeValue = androidHomeValue;
        // Check if we need to add an ANDROID_HOME value
        if (!tacoUtils.ProcessUtils.getProcess().env[AndroidSdkInstaller.ANDROID_HOME_NAME]) {
            exportAndroidHomeLine = util.format("%sexport %s=\"%s\"", os.EOL, AndroidSdkInstaller.ANDROID_HOME_NAME, androidHomeValue);
            // Modify the value of ANDROID_HOME for the running process
            process.env[AndroidSdkInstaller.ANDROID_HOME_NAME] = this.androidHomeValue;
        }
        else {
            var existingSdkHome = tacoUtils.ProcessUtils.getProcess().env[AndroidSdkInstaller.ANDROID_HOME_NAME];
            // Process the existing ANDROID_HOME to resolve to an absolute path, including processing ~ notation and environment variables
            existingSdkHome = path.resolve(utilHelper.expandEnvironmentVariables(existingSdkHome));
            if (existingSdkHome !== androidHomeValue) {
                // A conflicting ANDROID_HOME already exists, warn the user, but don't add our own ANDROID_HOME
                this.logger.logWarning(resources.getString("SystemVariableExistsDarwin", AndroidSdkInstaller.ANDROID_HOME_NAME, this.androidHomeValue));
                useShortPaths = false;
            }
        }
        // Check if we need to update PATH
        if (!installerUtils.pathContains(fullPathTools)) {
            addToPath += path.delimiter + (useShortPaths ? shortPathTools : fullPathTools);
            // Modify the value of PATH for the running process
            process.env["PATH"] += path.delimiter + fullPathTools;
        }
        if (!installerUtils.pathContains(fullPathPlatformTools)) {
            addToPath += path.delimiter + (useShortPaths ? shortPathPlatformTools : fullPathPlatformTools);
            // Modify the value of PATH for the running process
            process.env["PATH"] += path.delimiter + fullPathPlatformTools;
        }
        if (addToPath) {
            exportPathLine = util.format("%sexport PATH=\"$PATH%s\"", os.EOL, addToPath);
        }
        // Check if we need to update .bash_profile
        var bashProfilePath = path.join(tacoUtils.ProcessUtils.getProcess().env.HOME, ".bash_profile");
        var mustChown = !fs.existsSync(bashProfilePath);
        if (exportAndroidHomeLine || exportPathLine) {
            updateCommand = util.format("echo '%s# Android SDK%s%s' >> '%s'", os.EOL, exportAndroidHomeLine, exportPathLine, bashProfilePath);
        }
        // Perform the update if necessary
        if (updateCommand) {
            childProcess.exec(updateCommand, function (error, stdout, stderr) {
                if (error) {
                    _this.telemetry
                        .add("error.description", "ErrorOnChildProcess on updateVariablesDarwin", /*isPii*/ false)
                        .addError(error);
                    deferred.reject(error);
                }
                else {
                    // If .bash_profile didn't exist before, make sure the owner is the current user, not root
                    if (mustChown) {
                        fs.chownSync(bashProfilePath, parseInt(tacoUtils.ProcessUtils.getProcess().env.SUDO_UID, 10), parseInt(tacoUtils.ProcessUtils.getProcess().env.SUDO_GID, 10));
                    }
                    deferred.resolve({});
                }
            });
        }
        else {
            deferred.resolve({});
        }
        return deferred.promise;
    };
    AndroidSdkInstaller.prototype.postInstallDarwin = function () {
        var self = this;
        // We need to add execute permission to the android executable in order to run it
        return this.addExecutePermission(path.join(this.androidHomeValue, "tools", "android"))
            .then(function () {
            // We need to add execute permissions for the Gradle wrapper in order to build Android projects
            return self.addExecutePermission(path.join(self.androidHomeValue, "tools", "templates", "gradle", "wrapper", "gradlew"));
        });
    };
    AndroidSdkInstaller.prototype.downloadDefault = function () {
        this.installerArchive = path.join(InstallerBase.installerCache, "androidSdk", os.platform(), this.softwareVersion, path.basename(this.installerInfo.installSource));
        // Prepare expected archive file properties
        var expectedProperties = {
            bytes: this.installerInfo.bytes,
            sha1: this.installerInfo.sha1
        };
        // Prepare download options
        var options = {
            uri: this.installerInfo.installSource,
            method: "GET"
        };
        // Download the archive
        return installerUtils.downloadFile(options, this.installerArchive, expectedProperties);
    };
    AndroidSdkInstaller.prototype.installDefault = function () {
        // Make sure we have an install location
        if (!this.installDestination) {
            this.telemetry.add("error.description", "NeedInstallDestination on installDefault", /*isPii*/ false);
            return Q.reject(new Error(resources.getString("NeedInstallDestination")));
        }
        // Extract the archive
        var templateZip = new AdmZip(this.installerArchive);
        if (!fs.existsSync(this.installDestination)) {
            wrench.mkdirSyncRecursive(this.installDestination, 511); // 511 decimal is 0777 octal
        }
        templateZip.extractAllTo(this.installDestination);
        return Q.resolve({});
    };
    AndroidSdkInstaller.prototype.addExecutePermission = function (fileFullPath) {
        var _this = this;
        var deferred = Q.defer();
        var command = util.format("chmod a+x \"%s\"", fileFullPath);
        childProcess.exec(command, function (error, stdout, stderr) {
            if (error) {
                _this.telemetry
                    .add("error.description", "ErrorOnChildProcess on addExecutePermission", /*isPii*/ false)
                    .addError(error);
                deferred.reject(error);
            }
            else {
                deferred.resolve({});
            }
        });
        return deferred.promise;
    };
    AndroidSdkInstaller.ANDROID_HOME_NAME = "ANDROID_HOME";
    AndroidSdkInstaller.androidCommand = os.platform() === "win32" ? "android.bat" : "android";
    return AndroidSdkInstaller;
})(InstallerBase);
module.exports = AndroidSdkInstaller;
