// taco-cli/cli/tacoErrorCodes.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
/// <reference path="../../typings/node.d.ts" />
"use strict";
// Error Codes: 5000- 5999
var TacoErrorCode;
(function (TacoErrorCode) {
    // General errors 5000-5449
    // CordovaCmdNotFound = 5001,
    // CordovaCommandFailed = 5002,
    // CordovaCommandFailedWithError = 5003,
    TacoErrorCode[TacoErrorCode["NotInCordovaProject"] = 5004] = "NotInCordovaProject";
    TacoErrorCode[TacoErrorCode["CommandError"] = 5005] = "CommandError";
    TacoErrorCode[TacoErrorCode["UnimplementedAbstractMethod"] = 5006] = "UnimplementedAbstractMethod";
    TacoErrorCode[TacoErrorCode["ErrorTacoJsonMissingOrMalformed"] = 5007] = "ErrorTacoJsonMissingOrMalformed";
    TacoErrorCode[TacoErrorCode["ErrorNoPlatformsFound"] = 5008] = "ErrorNoPlatformsFound";
    TacoErrorCode[TacoErrorCode["CommandBuildTacoSettingsNotFound"] = 5009] = "CommandBuildTacoSettingsNotFound";
    TacoErrorCode[TacoErrorCode["TacoSettingsFileDoesNotExist"] = 5010] = "TacoSettingsFileDoesNotExist";
    TacoErrorCode[TacoErrorCode["CommandBuildInvalidPlatformLocation"] = 5011] = "CommandBuildInvalidPlatformLocation";
    TacoErrorCode[TacoErrorCode["ErrorOperationCancelled"] = 5012] = "ErrorOperationCancelled";
    TacoErrorCode[TacoErrorCode["ErrorInvalidVersion"] = 5013] = "ErrorInvalidVersion";
    TacoErrorCode[TacoErrorCode["ErrorReadingPackageVersions"] = 5014] = "ErrorReadingPackageVersions";
    // CordovaCommandUnhandledException = 5015,
    // InvalidCordovaWithNode5 = 5016,
    // Errors to do with Create 5450-5499
    TacoErrorCode[TacoErrorCode["CommandCreateGitCloneError"] = 5451] = "CommandCreateGitCloneError";
    TacoErrorCode[TacoErrorCode["CommandCreateInvalidPath"] = 5452] = "CommandCreateInvalidPath";
    TacoErrorCode[TacoErrorCode["CommandCreateNoGit"] = 5453] = "CommandCreateNoGit";
    // CommandCreateTacoJsonFileCreationError = 5454,
    // CommandCreateTacoJsonFileWriteError = 5455,
    TacoErrorCode[TacoErrorCode["CommandCreateTemplatesUnavailable"] = 5456] = "CommandCreateTemplatesUnavailable";
    // Invalid Arguments 5500 - 5599
    // General invalid arguments 5500 - 5509
    TacoErrorCode[TacoErrorCode["ErrorIncompatibleOptions"] = 5501] = "ErrorIncompatibleOptions";
    // Invalid arguments for remote 5510 - 5529
    TacoErrorCode[TacoErrorCode["CommandRemotePlatformNotKnown"] = 5511] = "CommandRemotePlatformNotKnown";
    TacoErrorCode[TacoErrorCode["CommandRemoteDeleteNeedsPlatform"] = 5512] = "CommandRemoteDeleteNeedsPlatform";
    TacoErrorCode[TacoErrorCode["CommandRemoteDeletePlatformNotAdded"] = 5513] = "CommandRemoteDeletePlatformNotAdded";
    TacoErrorCode[TacoErrorCode["CommandRemoteInvalidPin"] = 5514] = "CommandRemoteInvalidPin";
    TacoErrorCode[TacoErrorCode["CommandRemoteInvalidPort"] = 5515] = "CommandRemoteInvalidPort";
    // Invalid arguments for create 5530 - 5549
    TacoErrorCode[TacoErrorCode["CommandCreateNoPath"] = 5531] = "CommandCreateNoPath";
    TacoErrorCode[TacoErrorCode["CommandCreateNotBothCordovaCliAndKit"] = 5532] = "CommandCreateNotBothCordovaCliAndKit";
    TacoErrorCode[TacoErrorCode["CommandCreateNotBothTemplateAndCordovaCli"] = 5533] = "CommandCreateNotBothTemplateAndCordovaCli";
    TacoErrorCode[TacoErrorCode["CommandCreateNotTemplateIfCustomWww"] = 5534] = "CommandCreateNotTemplateIfCustomWww";
    // CommandCreateOnlyLocalCustomWww = 5535,
    TacoErrorCode[TacoErrorCode["CommandCreatePathNotEmpty"] = 5536] = "CommandCreatePathNotEmpty";
    // Invalid arguments for kits 5550 - 5569
    TacoErrorCode[TacoErrorCode["CommandKitCliOrKitShouldBeSpecified"] = 5551] = "CommandKitCliOrKitShouldBeSpecified";
    TacoErrorCode[TacoErrorCode["CommandKitProjectUsesSameKit"] = 5552] = "CommandKitProjectUsesSameKit";
    TacoErrorCode[TacoErrorCode["CommandKitProjectUsesSameCordovaCli"] = 5553] = "CommandKitProjectUsesSameCordovaCli";
    TacoErrorCode[TacoErrorCode["ErrorInvalidJsonFilePath"] = 5554] = "ErrorInvalidJsonFilePath";
    TacoErrorCode[TacoErrorCode["ErrorInvalidPath"] = 5555] = "ErrorInvalidPath";
    TacoErrorCode[TacoErrorCode["ErrorNoPluginOrPlatformSpecified"] = 5556] = "ErrorNoPluginOrPlatformSpecified";
    // Errors to do with remote 5600-5699
    TacoErrorCode[TacoErrorCode["CommandRemoteCantFindRemoteMount"] = 5600] = "CommandRemoteCantFindRemoteMount";
    TacoErrorCode[TacoErrorCode["CommandRemoteConnectionRefused"] = 5601] = "CommandRemoteConnectionRefused";
    TacoErrorCode[TacoErrorCode["CommandRemoteNotfound"] = 5602] = "CommandRemoteNotfound";
    TacoErrorCode[TacoErrorCode["CommandRemoteRejectedPin"] = 5603] = "CommandRemoteRejectedPin";
    TacoErrorCode[TacoErrorCode["CommandRemoteTimedout"] = 5604] = "CommandRemoteTimedout";
    TacoErrorCode[TacoErrorCode["ErrorDownloadingRemoteBuild"] = 5605] = "ErrorDownloadingRemoteBuild";
    TacoErrorCode[TacoErrorCode["ErrorHttpGet"] = 5606] = "ErrorHttpGet";
    TacoErrorCode[TacoErrorCode["ErrorUploadingRemoteBuild"] = 5607] = "ErrorUploadingRemoteBuild";
    TacoErrorCode[TacoErrorCode["ErrorDuringRemoteBuildSubmission"] = 5608] = "ErrorDuringRemoteBuildSubmission";
    TacoErrorCode[TacoErrorCode["ErrorCertificateLoad"] = 5609] = "ErrorCertificateLoad";
    TacoErrorCode[TacoErrorCode["ErrorCertificatePathChmod"] = 5610] = "ErrorCertificatePathChmod";
    TacoErrorCode[TacoErrorCode["ErrorCertificateSave"] = 5611] = "ErrorCertificateSave";
    TacoErrorCode[TacoErrorCode["ErrorCertificateSaveToPath"] = 5612] = "ErrorCertificateSaveToPath";
    TacoErrorCode[TacoErrorCode["ErrorCertificateSaveWithErrorCode"] = 5613] = "ErrorCertificateSaveWithErrorCode";
    TacoErrorCode[TacoErrorCode["UnsupportedHostPlatform"] = 5614] = "UnsupportedHostPlatform";
    TacoErrorCode[TacoErrorCode["GetCertificateFailed"] = 5615] = "GetCertificateFailed";
    TacoErrorCode[TacoErrorCode["HttpGetFailed"] = 5616] = "HttpGetFailed";
    TacoErrorCode[TacoErrorCode["InvalidBuildSubmission400"] = 5617] = "InvalidBuildSubmission400";
    TacoErrorCode[TacoErrorCode["InvalidRemoteBuild"] = 5618] = "InvalidRemoteBuild";
    TacoErrorCode[TacoErrorCode["InvalidRemoteBuildClientCert"] = 5619] = "InvalidRemoteBuildClientCert";
    TacoErrorCode[TacoErrorCode["InvalidRemoteBuildUrl"] = 5620] = "InvalidRemoteBuildUrl";
    TacoErrorCode[TacoErrorCode["NoCertificateFound"] = 5621] = "NoCertificateFound";
    TacoErrorCode[TacoErrorCode["NoRemoteBuildIdFound"] = 5622] = "NoRemoteBuildIdFound";
    TacoErrorCode[TacoErrorCode["RemoteBuildError"] = 5623] = "RemoteBuildError";
    TacoErrorCode[TacoErrorCode["RemoteBuildHostNotFound"] = 5624] = "RemoteBuildHostNotFound";
    TacoErrorCode[TacoErrorCode["RemoteBuildNoConnection"] = 5625] = "RemoteBuildNoConnection";
    TacoErrorCode[TacoErrorCode["RemoteBuildNonSslConnectionReset"] = 5626] = "RemoteBuildNonSslConnectionReset";
    TacoErrorCode[TacoErrorCode["RemoteBuildSslConnectionReset"] = 5627] = "RemoteBuildSslConnectionReset";
    TacoErrorCode[TacoErrorCode["RemoteBuildStatusPollFailed"] = 5628] = "RemoteBuildStatusPollFailed";
    TacoErrorCode[TacoErrorCode["RemoteBuildUnsuccessful"] = 5629] = "RemoteBuildUnsuccessful";
    TacoErrorCode[TacoErrorCode["RemoteBuildUnsupportedPlatform"] = 5630] = "RemoteBuildUnsupportedPlatform";
    TacoErrorCode[TacoErrorCode["ErrorPatchCreation"] = 5631] = "ErrorPatchCreation";
    // Errors related to install-reqs 5700-5749
    TacoErrorCode[TacoErrorCode["InstallReqsNotSupported"] = 5700] = "InstallReqsNotSupported";
    TacoErrorCode[TacoErrorCode["CommandInstallNoPlatformsAdded"] = 5701] = "CommandInstallNoPlatformsAdded";
    TacoErrorCode[TacoErrorCode["CommandInstallNoPlatformsFolder"] = 5702] = "CommandInstallNoPlatformsFolder";
})(TacoErrorCode || (TacoErrorCode = {}));
module.exports = TacoErrorCode;
