// taco-dependency-installer/elevatedInstallerProtocol.js
/**
 *******************************************************
 *                                                     *
 *   Copyright (C) Microsoft. All rights reserved.     *
 *                                                     *
 *******************************************************
 */
/// <reference path="../typings/node.d.ts" />
/// <reference path="../typings/q.d.ts" />
"use strict";
var ElevatedInstallerProtocol;
(function (ElevatedInstallerProtocol) {
    /**
     * These exit codes are arbitrary (except for the success one). They were chosen because they seemed untaken by Powershell and Node. We need such error codes to differentiate between Powershell and Node exit
     * codes, as well as the possible outcomes of our child-process tree that has 3 levels of depth.
     *
     * Note: on Unix, return codes are limited to 1 byte, so these codes need to stay below 256.
     *
     * Powershell exit codes (just 0 and 1): http://blogs.msdn.com/b/powershell/archive/2006/10/14/windows-powershell-exit-codes.aspx (article is old, but I couldn't find evidence that it has changed)
     * Node.js exit codes (0 through 10, 12 and 128): https://github.com/nodejs/node-v0.x-archive/blob/master/doc/api/process.markdown#exit-codes
     */
    (function (ExitCode) {
        ExitCode[ExitCode["Success"] = 0] = "Success";
        ExitCode[ExitCode["CompletedWithErrors"] = 12] = "CompletedWithErrors";
        ExitCode[ExitCode["NoAdminRights"] = 33] = "NoAdminRights";
        ExitCode[ExitCode["CouldNotConnect"] = 37] = "CouldNotConnect";
        ExitCode[ExitCode["RefusedPrompt"] = 42] = "RefusedPrompt";
        ExitCode[ExitCode["FatalError"] = 108] = "FatalError";
    })(ElevatedInstallerProtocol.ExitCode || (ElevatedInstallerProtocol.ExitCode = {}));
    var ExitCode = ElevatedInstallerProtocol.ExitCode;
    (function (DataType) {
        DataType[DataType["Error"] = 0] = "Error";
        DataType[DataType["Log"] = 1] = "Log";
        DataType[DataType["Prompt"] = 2] = "Prompt";
        DataType[DataType["Warning"] = 3] = "Warning";
    })(ElevatedInstallerProtocol.DataType || (ElevatedInstallerProtocol.DataType = {}));
    var DataType = ElevatedInstallerProtocol.DataType;
})(ElevatedInstallerProtocol || (ElevatedInstallerProtocol = {}));
module.exports = ElevatedInstallerProtocol;
