// taco-dependency-installer/tacoErrorCodes.js
/**
 *******************************************************
 *                                                     *
 *   Copyright (C) Microsoft. All rights reserved.     *
 *                                                     *
 *******************************************************
 */
/// <reference path="../typings/node.d.ts" />
"use strict";
var TacoDependencyInstaller;
(function (TacoDependencyInstaller) {
    // Error Codes: 2100-2499
    (function (TacoErrorCode) {
        TacoErrorCode[TacoErrorCode["AbstractMethod"] = 2101] = "AbstractMethod";
        TacoErrorCode[TacoErrorCode["CouldNotConnect"] = 2102] = "CouldNotConnect";
        TacoErrorCode[TacoErrorCode["ErrorCreatingInstallConfig"] = 2103] = "ErrorCreatingInstallConfig";
        TacoErrorCode[TacoErrorCode["ErrorDeletingInstallConfig"] = 2104] = "ErrorDeletingInstallConfig";
        TacoErrorCode[TacoErrorCode["FatalError"] = 2105] = "FatalError";
        TacoErrorCode[TacoErrorCode["FileCorruptError"] = 2106] = "FileCorruptError";
        TacoErrorCode[TacoErrorCode["FileNotFound"] = 2107] = "FileNotFound";
        TacoErrorCode[TacoErrorCode["InstallationErrors"] = 2108] = "InstallationErrors";
        TacoErrorCode[TacoErrorCode["NoAdminRights"] = 2110] = "NoAdminRights";
        TacoErrorCode[TacoErrorCode["NoPowershell"] = 2111] = "NoPowershell";
        TacoErrorCode[TacoErrorCode["NoValidInstallOrder"] = 2112] = "NoValidInstallOrder";
        TacoErrorCode[TacoErrorCode["UnknownExitCode"] = 2113] = "UnknownExitCode";
        TacoErrorCode[TacoErrorCode["UnsupportedPlatform"] = 2114] = "UnsupportedPlatform";
        TacoErrorCode[TacoErrorCode["TacoKitsExceptionTypescriptNotSupported"] = 2115] = "TacoKitsExceptionTypescriptNotSupported";
    })(TacoDependencyInstaller.TacoErrorCode || (TacoDependencyInstaller.TacoErrorCode = {}));
    var TacoErrorCode = TacoDependencyInstaller.TacoErrorCode;
})(TacoDependencyInstaller || (TacoDependencyInstaller = {}));
module.exports = TacoDependencyInstaller;
